File: Javascript/quick-sort.js
// Define the quickSort function, along with the partition function that it will call

function quickSort(array, leftPointer = 0, rightPointer = array.length - 1) {
  if (leftPointer < rightPointer) {
    var pivot = Math.floor(Math.random() * (rightPointer - leftPointer) + leftPointer);
    var breakPoint = partition(array, leftPointer, rightPointer, pivot)
    quickSort(array, leftPointer, breakPoint - 1);
    quickSort(array, breakPoint + 1, rightPointer);
  }
  return array;
}

function partition(array, leftPointer, rightPointer, pivot) {
  pivotValue = array[pivot];
  array[pivot] = array[rightPointer];
  array[rightPointer] = pivotValue;

  storageMarker = leftPointer;

  for (var i = leftPointer; i < rightPointer; i++) {
    if (array[i] < array[rightPointer]) {
      lesserValue = array[i];
      array[i] = array[storageMarker];
      array[storageMarker] = lesserValue;
      storageMarker += 1;
    }
  }

  storageValue = array[storageMarker];
  array[storageMarker] = array[rightPointer];
  array[rightPointer] = storageValue;

  return storageMarker;
}

// Define an array

var array = [7, 6, 4, 89, 67, 3, 53, 1, 2, 76, 7]

// Call the quickSort function and log the results to the console

console.log(quickSort(array));

File: Java/src/QuickSort.java
import java.util.Arrays;

public class QuickSort {

	public static int[] quickSort(int[] arr, int leftCursor, int rightCursor) {
		if (leftCursor < rightCursor) {
			int pivot = leftCursor + (int)(Math.floor((Math.random()) * (rightCursor - leftCursor + 1)));
			int breakpoint = partition(arr, leftCursor, rightCursor, pivot);
			quickSort(arr, leftCursor, breakpoint - 1);
			quickSort(arr, breakpoint + 1, rightCursor);
		}
		return arr;
	}
	
	public static int partition(int[] arr, int leftCursor, int rightCursor, int pivot) {
		int pivotValue = arr[pivot];
		arr[pivot] = arr[rightCursor];
		arr[rightCursor] = pivotValue;
		
		int storageMarker = leftCursor;
		
		for (int i = leftCursor; i < rightCursor; i++) {
			if (arr[i] < pivotValue) {
				int lesserValue = arr[i];
				arr[i] = arr[storageMarker];
				arr[storageMarker] = lesserValue;
				storageMarker += 1;
			}
		}
		
		int storageValue = arr[storageMarker];
		arr[storageMarker] = arr[rightCursor];
		arr[rightCursor] = storageValue;
		
		return storageMarker;
	}
	
	public static void main(String[] args) {
		int[] arr = {7, 4, 3, 78, 6, 45, 2, 4, 32, 10};
		System.out.println("Unsorted array:");
		System.out.println(Arrays.toString(arr));
		System.out.println("Sorted array:");
		System.out.println(Arrays.toString(quickSort(arr, 0, arr.length - 1)));
	}

}
